src/components/Form.tsx

```diff
@@ -0,0 +1,143 @@
+import React, { useContext, useEffect, useState } from "react";
+import { useHistory } from "react-router-dom";
+import styled from "styled-components";
+import { createRecipe, deleteRecipe, fetchRecipes, modifyRecipe } from "../api";
+import { RecipeContext } from "../context";
+import { Recipe } from "../models";
+import { ConfirmButton, DeleteButton } from "./Button";
+import { IconSelector } from "./IconSelector";
+import { IngredientForm } from "./IngredientForm";
+
+export const Label = styled.label`
+  display: block;
+  margin-bottom: 0.5rem;
+  margin-top: 1.5rem;
+`;
+
+export const Input = styled.input`
+  font-size: 1rem;
+  font-family: inherit;
+  border: none;
+  border-radius: 8px;
+  padding: 0.5rem 0.75rem;
+  box-shadow: 0 0px 1px hsla(0, 0%, 0%, 0.2), 0 1px 2px hsla(0, 0%, 0%, 0.2);
+  background-color: white;
+  line-height: 1.5;
+  margin: 0;
+`;
+
+type Props = {
+  recipe?: Recipe | null;
+};
+
+export function RecipeForm({ recipe }: Props) {
+  const { recipes, setRecipes } = useContext(RecipeContext);
+  const history = useHistory();
+
+  const [recipeState, setRecipeState] = useState<Recipe>(
+    recipe ||
+      ({
+        name: "",
+        description: "",
+        icon: "🍔",
+        ingredients: [],
+      } as Recipe)
+  );
+
+  useEffect(() => {
```

Comment:
I think we should be able to remove this use effect, since it's already setting the recipe as the initial state of the `useState`

Reply:
Due to the way React router works, this useEffect is needed when you change the `id` parameter in the URL. React router does not rerender the component, it just updates the state, which does not visibly update the page, unless there is a useEffect on the page.

I will add a comment explaining this as it is not obvious.